sycl/doc/extensions/experimental/sycl_ext_oneapi_enqueue_native_command.asciidoc

```diff
@@ -0,0 +1,240 @@
+implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== Additions to handler class
+
+This extension adds the following new member function to the SYCL `handler`
+class:
+
+```c++
+class handler {
+  template <typename Func>
+  void ext_oneapi_enqueue_native_command(Func&& interopCallable);
+};
+```
+
+_Constraints_: The `Func` must a C++ callable object which takes a single
+parameter of type `interop_handle`.
+
+_Effects_: The `interopCallable` object is called exactly once, and this call
+may be made asynchronously even after the calling thread returns from
+`ext_oneapi_enqueue_native_command`.
+
+The call to `interopCallable` may submit one or more asynchronous tasks to the
+native backend object obtained from `interop_handle::get_native_queue`, and
+these tasks become encapsulated in a SYCL command that is added to the queue.
+If the enclosing command group has any dependencies, these dependencies will be
+propagated to the native API by the SYCL implementation during the command
+submission, meaning no further dependency analysis is needed within the user
+submitted `interopCallable` in the native API. This happens, for example, if
+the command group calls `handler::depends_on` or if it constructs an accessor.
+Note, however, that the SYCL command is associated with the _asynchronous tasks
+submitted by_ `interopCallable`, not the call to `interopCallable`. The call to
+`interopCallable` may happen even before the dependencies are satisfied.
+
+The SYCL command described above completes once all of the native asynchronous
+tasks it contains have completed.
+
+The call to interopCallable must not submit any synchronous tasks to the native
+backend object, and it must not block waiting for any tasks to complete. The
+call also must not add tasks to backend objects that underly any other queue,
+aside from the queue that is associated with this handler. If it does any of
+these things, the behavior is undefined.
+
+== Examples
+
+=== Correct usage
+
+This example demonstrates how to use this extension to enqueue asynchronous
+native tasks on the HIP backend.
+
+```c++
+sycl::queue q;
+q.submit([&](sycl::handler &cgh) {
+    sycl::accessor acc{buf, cgh};
+
+    cgh.ext_oneapi_enqueue_native_command([=](sycl::interop_handle &h) {
+      // Can extract device pointers from accessors
+      void *native_mem = h.get_native_mem<sycl::backend::hip>(acc);
+      // Can extract stream
+      hipStream_t stream = h.get_native_queue<sycl::backend::hip>();
+
+      // Can enqueue arbitrary backend operations. This could also be a kernel
+      // launch or call to a library that enqueues operations on the stream etc
+      hipMemcpyAsync(target_ptr, native_mem, test_size * sizeof(int),
+                      hipMemcpyDeviceToHost, stream);
+    });
+  });
+q.wait();
+```
+
+=== Incorrect usage
```

Comment:
@Pennycook do you have any comments about adding examples of incorrect code to our extension specification?

Reply:
My recommendation is that we don't do this.

I worry that there are a lot of people that use Ctrl+F to jump to parts of specifications, who might miss the header declaring that this is an "Incorrect" example.  Putting this incorrect example onto GitHub is also likely to put it in the training set of LLMs, increasing the likelihood of things like Copilot generating incorrect SYCL code.

I think it would be more effective to combine the information from the two examples, something like this:

```c++
sycl::queue q;
q.submit([&](sycl::handler &cgh) {
    sycl::accessor acc{buf, cgh};

    cgh.ext_oneapi_enqueue_native_command([=](sycl::interop_handle &h) {
      // Can extract device pointers from accessors
      void *native_mem = h.get_native_mem<sycl::backend::hip>(acc);
      // Can extract stream
      hipStream_t stream = h.get_native_queue<sycl::backend::hip>();

      // Can enqueue arbitrary *asynchronous* backend operations.
      // This could be a kernel launch or call to a library that enqueues operations on the stream, etc.
      //
      // Important: Enqueuing a *synchronous* backend operation results in undefined behavior.
      hipMemcpyAsync(target_ptr, native_mem, test_size * sizeof(int),
                      hipMemcpyDeviceToHost, stream);
    });
  });
q.wait();
```

As an added advantage, readers only have to read one example to get the same amount of information.

Reply:
I agree. Thanks, have updated.